File: js/packages/teams-ai/src/prompts/FunctionResponseMessage.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { Message } from './Message';
import { PromptFunctions } from './PromptFunctions';
import { RenderedPromptSection } from './PromptSection';
import { PromptSectionBase } from './PromptSectionBase';
import { Utilities } from '../Utilities';
import { TurnContext } from 'botbuilder';
import { Tokenizer } from '../tokenizers';
import { Memory } from '../MemoryFork';

/**
 * Message containing the response to a function call.
 */
export class FunctionResponseMessage extends PromptSectionBase {
    private _text: string = '';
    private _length: number = -1;

    public readonly name: string;
    public readonly response: any;

    /**
     * Creates a new 'FunctionResponseMessage' instance.
     * @param {string} name - Name of the function that was called.
     * @param {any} response - The response returned by the called function.
     * @param {number} tokens - Optional. Sizing strategy for this section. Defaults to `auto`.
     * @param {string} functionPrefix - Optional. Prefix to use for function messages when rendering as text. Defaults to `user: ` to simulate the response coming from the user.
     */
    public constructor(name: string, response: any, tokens: number = -1, functionPrefix: string = 'user: ') {
        super(tokens, true, '\n', functionPrefix);
        this.name = name;
        this.response = response;
    }

    /**
     * @private
     * @param {TurnContext} context - Context for the current turn of conversation.
     * @param {Memory} memory - Memory to use for rendering.
     * @param {PromptFunctions} functions - Prompt functions to use for rendering.
     * @param {Tokenizer} tokenizer - Tokenizer to use for encoding text.
     * @param {number} maxTokens - Maximum number of tokens allowed.
     * @returns {Promise<RenderedPromptSection<Message[]>>} Rendered prompt section as a string.
     */
    public async renderAsMessages(
        context: TurnContext,
        memory: Memory,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        maxTokens: number
    ): Promise<RenderedPromptSection<Message[]>> {
        // Calculate and cache response text and length
        if (this._length < 0) {
            this._text = Utilities.toString(tokenizer, this.response);
            this._length = tokenizer.encode(this.name).length + tokenizer.encode(this._text).length;
        }

        // Return output
        return this.returnMessages(
            [{ role: 'function', name: this.name, content: this._text }],
            this._length,
            tokenizer,
            maxTokens
        );
    }
}

File: python/packages/ai/teams/ai/prompts/function_response_message.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import Any, List

from botbuilder.core import TurnContext

from ...state import MemoryBase
from ..tokenizers import Tokenizer
from ..utilities import to_string
from .message import Message
from .prompt_functions import PromptFunctions
from .rendered_prompt_section import RenderedPromptSection
from .sections.prompt_section_base import PromptSectionBase


class FunctionResponseMessage(PromptSectionBase):
    """
    Message containing the response to a function call.
    """

    _text: str
    _length: int
    _name: str
    _response: Any

    def __init__(
        self, name: str, response: Any, tokens: float = -1, function_prefix: str = "user: "
    ):
        """
        Creates a new `FunctionResponseMessage` instance.

        Args:
            name (str): Name of the function that was called.
            response (Any): The response returned by the called function.
            tokens (int, optional): Sizing strategy for this section. Defaults to `auto`.
            function_prefix (str, optional): Prefix to use for function messages when
              rendering as text. Defaults to `user: `  to simulate the response
              coming from the user..

        """
        super().__init__(tokens, True, "\n", function_prefix)
        self._text = ""
        self._length = -1
        self._name = name
        self._response = response

    @property
    def name(self) -> str:
        """
        Name of the function that was called.
        """
        return self._name

    @property
    def response(self) -> Any:
        """
        The response returned by the called function.
        """
        return self._response

    async def render_as_messages(
        self,
        context: TurnContext,
        memory: MemoryBase,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        max_tokens: int,
    ) -> RenderedPromptSection[List[Message]]:
        """
        Renders the section as a list of messages.

        Args:
            context (TurnContext): Context for the current turn of conversation with the user.
            memory (MemoryBase): An interface for accessing state values.
            functions (PromptFunctions): Registry of functions that can be used by the section.
            tokenizer (Tokenizer): Tokenizer to use when rendering the section.
            max_tokens (int): Maximum number of tokens allowed to be rendered.

        Returns:
            RenderedPromptSection[List[Message]]: The rendered prompt section as a list of messages.
        """
        # Calculate and cache response text and length
        if self._length < 0:
            self._text = to_string(tokenizer, self.response)
            self._length = len(tokenizer.encode(self.name)) + len(tokenizer.encode(self._text))

        # Return output
        return self._return_messages(
            [Message("function", name=self.name, content=self._text)],
            self._length,
            tokenizer,
            max_tokens,
        )
